rust/flexbuffers/src/flexbuffer.rs

```diff
@@ -0,0 +1,89 @@
+use std::ops::{RangeBounds, Rem, Bound};
+use std::slice::SliceIndex;
+use std::fmt::Debug;
+
+use crate::reader::ReadLE;
+
+pub trait FlexBuffer: Debug + PartialEq + IntoIterator + AsRef<[u8]> + Default + Clone {
+    /// Gets the length of the current buffer.
+    fn len(&self) -> usize;
```

Comment:
I think all these functions don't need to be part of this trait as we can always get them from `AsRef<[u8]>`

Reply:
That's a great point! I'll remove the extra functions. 

One big reason on keeping the `slice` function is so that other entities (specifically the `Blob` struct) can then keep using the generic `Buffer` type without worrying about lifetimes. 

If there's a way around this issue though that would be helpful, since then there wouldn't even need to be a `FlexBuffer` / `Buffer` trait, you could just do:

```rust
struct Reader<B: AsRef<u8>> {}
```

Which would allow for automatic 3rd party library support in many cases. Would it be a big deal if a `Blob` object referred to a lifetime bound function? It would require some lifetime wrangling for the `get_blob` function.

Reply:
> One big reason on keeping the slice function is so that other entities (specifically the Blob struct) can then keep using the generic Buffer type without worrying about lifetimes.

Ok that makes sense. 

>  Would it be a big deal if a Blob object referred to a lifetime bound function? 

No, I think that'll break borrowed deserialization.

https://github.com/google/flatbuffers/blob/6f3e45eca1fde7a68cb72fd4499a3647f719c9db/rust/flexbuffers/src/reader/de.rs#L178

I realize that `AsRef<[u8]>` isn't exactly what we're looking for. We want to support `&Buffer<'de> -> &'de[u8]` as well as `&Buffer -> &[u8]` and the former lifetime is incompatible with `AsRef<[u8]>` so indeed you need your own slice method to do the lifetimes properly. If that's the case, maybe you should remove `AsRef<[u8]>` to make the trait more minimal?

Reply:
I've made it so `Blob` just points to a buffer `&[u8]` so that anything that has support for `Clone` and `AsRef<[u8]>` will be supported.

Reply:
Oh just saw this, going to review how to make deserialization lifetimes work correctly. 

For now I was using `PhantomData` which took a pointer with a lifetime as a type. This way the underlying `[u8]` will always be assured to be present throughout deserialization. 

```rust
pub struct ReaderDeserializer<'de, B> {
    inner: Reader<B>,
    lifetime: PhantomData<&'de B>
}
```

I think this might help eliminate the need for `slice`, since this would ensure that any kind of deserialization done through `serde` would always know the given bytes exist. That way users could implement a Reader directly, it would be a breaking change. I'll try to do some testing to see if I can potentially avoid unsafe code usage in general.

Reply:
I'll also see if removing `AsRef<[u8]>` is the way to go. 

Alternatively I was thinking of just implementing the serde `Deserializer` itself just for the `&'de [u8]` type.